File: 1048-longest-string-chain/1048-longest-string-chain.java
class Solution {
    public int longestStrChain(String[] words) {
        Arrays.sort(words,(a,b) ->a.length()-b.length());
        int res = 0;
        Map<String,Integer> memo = new HashMap<>();
        for(int i=0;i<words.length;i++){
            memo.put(words[i],1);
            for(int j=0;j<words[i].length();j++){
                StringBuilder cur = new StringBuilder(words[i]);
                String next = cur.deleteCharAt(j).toString();
                if(memo.containsKey(next)){
                    memo.put(words[i],Math.max(memo.get(words[i]),memo.get(next)+1));
                }
            }
            res = Math.max(memo.get(words[i]),res);
        }
        return res;
    }
}

File: 1048-longest-string-chain/1048-longest-string-chain.cpp
class Solution {
public:
    bool check(string &a, string &b){
        int m = a.size();
        int n = b.size();
        if(m!=n+1){
            return false;
        }
        int first =0;
        int second =0;
        while(first<m){
            if(a[first]==b[second]){
                first++;
                second++;
            }
            else{
                first++;
            }
        }
        if(first == m && second ==n){
            return true;
        }
        return false;
    }

    bool static comp(string& s1, string& s2){
        return s1.size() < s2.size();
    }

    int longestStrChain(vector<string> &arr){
        int n=arr.size();
        sort(arr.begin(),arr.end(),comp);
        vector<int> dp(n,1);
        int maxi=1;
        for(int i=1;i<n;i++){
            for(int j=0;j<i;j++){
                if(check(arr[i],arr[j])){
                    dp[i] = max(1+dp[j],dp[i]);
                }
            }
            if(dp[i]>maxi){
                maxi = max(maxi,dp[i]);
            }
        }
        return maxi;
    }
};

